lib/gameState/CGameState.h

```diff
@@ -39,12 +39,67 @@ struct SThievesGuildInfo;
 class CRandomGenerator;
 class GameSettings;
 
-struct UpgradeInfo
+class UpgradeInfo
 {
+public:
+	UpgradeInfo() = delete;
+	UpgradeInfo(CreatureID base)
+		: oldID(base), isAvailable(true)
+	{ }
+
 	CreatureID oldID; //creature to be upgraded
-	std::vector<CreatureID> newID; //possible upgrades
-	std::vector<ResourceSet> cost; // cost[upgrade_serial] -> set of pairs<resource_ID,resource_amount>; cost is for single unit (not entire stack)
-	UpgradeInfo(){oldID = CreatureID::NONE;};
+
+	const auto & getAvailableUpgrades() const
```

Comment:
Not sure about auto as return type since it reduces readability a bit.
Previously I only needed to make a quick look on method signature to determine its return type. Now I need to also lookup what `upgradesIDs` is. Trivial in this method, but not so trivial in some of the other methods in this class.

Sure, IDE will detect return type, and show it in a tooltip, but when looking through list of methods it is way easier to locate method you're looking for with explicit return types.

So proposition: replace all deduced / auto return types with concrete values